framework/db/SqlToken.php

```diff
@@ -0,0 +1,301 @@
+     * It is implicitly called when you use something like `unset($token[$offset])`.
+     * @param int $offset child token offset.
+     */
+    public function offsetUnset($offset)
+    {
+        $offset = $this->calculateOffset($offset);
+        if (isset($this->_children[$offset])) {
+            array_splice($this->_children, $offset, 1);
+        }
+        $this->updateCollectionOffsets();
+    }
+
+    /**
+     * Returns child tokens.
+     * @return SqlToken[] child tokens.
+     */
+    public function getChildren()
+    {
+        return $this->_children;
+    }
+
+    /**
+     * Sets a list of child tokens.
+     * @param SqlToken[] $children child tokens.
+     */
+    public function setChildren($children)
+    {
+        $this->_children = [];
+        foreach ($children as $child) {
+            $child->parent = $this;
+            $this->_children[] = $child;
+        }
+        $this->updateCollectionOffsets();
+    }
+
+    /**
+     * Returns whether the token represents a collection of tokens.
+     * @return bool whether the token represents a collection of tokens.
+     */
+    public function getIsCollection()
+    {
+        return in_array($this->type, [
+            self::TYPE_CODE,
+            self::TYPE_STATEMENT,
+            self::TYPE_PARENTHESIS,
+        ], true);
+    }
+
+    /**
+     * Returns whether the token represents a collection of tokens and has non-zero number of children.
+     * @return bool whether the token has children.
+     */
+    public function getHasChildren()
+    {
+        return $this->getIsCollection() && !empty($this->_children);
+    }
+
+    /**
+     * Returns the SQL code representing the token.
+     * @return string SQL code.
+     */
+    public function getSql()
+    {
+        $code = $this;
+        while ($code->parent !== null) {
+            $code = $code->parent;
+        }
+        return mb_substr($code->content, $this->startOffset, $this->endOffset - $this->startOffset, 'UTF-8');
+    }
+
+    /**
+     * Returns whether this token (including its children) matches the specified "pattern" SQL code.
+     * Currently this implementation uses [[\yii\db\sqlite\SqlTokenizer]] to process a pattern SQL code.
+     * @param string $pattern SQL code. Everything supported in SQLite is supported here as well.
+     * In addition, `any` keyword is supported which is treated as any number of keywords, identifiers, whitespaces.
+     * @param int $offset token children offset to start lookup with.
+     * @param int|null $firstMatchIndex token children offset where a successful match begins.
+     * @param int|null $lastMatchIndex token children offset where a successful match ends.
+     * @return bool whether this token matches the pattern SQL code.
+     */
+    public function matches($pattern, $offset = 0, &$firstMatchIndex = null, &$lastMatchIndex = null)
+    {
+        $patternToken = (new \yii\db\sqlite\SqlTokenizer($pattern))->tokenize();
```

Comment:
SQLite's one is only implemented ATM so it's generic for now. We can change it later.